src/security/context.cpp

```diff
@@ -870,94 +1008,524 @@ ngx_int_t Context::do_request_body_filter(ngx_http_request_t &request,
+      ngx_log_error(
+          NGX_LOG_DEBUG, pool.log, 0,
+          "buffer_chain link: "
+          "t:%d m: %d mmap: %d, r:%d f:%d fl:%d lb=%d s:%d %p %p-%p %p %O-%O",
+          buf.temporary, buf.memory, buf.mmap, buf.recycled, buf.in_file,
+          buf.flush, buf.last_buf, buf.sync, buf.start, buf.pos, buf.last,
+          buf.file, buf.file_pos, buf.file_last);
+    }
+  }
+
+  if (in && filter_ctx.found_last) {
+    ngx_log_error(
+        NGX_LOG_NOTICE, pool.log, 0,
+        "given buffer after having already received one with ->last_buf");
+    return NGX_ERROR;
+  }
+
   for (auto *in_ch = in; in_ch; in_ch = in_ch->next) {
     ngx_chain_t *new_ch = ngx_alloc_chain_link(&pool);  // uninitialized
     if (!new_ch) {
       return NGX_ERROR;
     }
 
     auto *buf = in_ch->buf;
-    auto size = buf->last - buf->pos;
-    if (consume || buf->recycled) {  // copy the buffer and consume the original
-      ngx_buf_t *new_buf = ngx_create_temp_buf(&pool, size);
-      if (!new_buf) {
-        return NGX_ERROR;
+    size_t size;
+    if (consume) {  // copy the buffer and consume the original
+      ngx_buf_t *new_buf;
+      if (!buf->in_file) {
+        size = buf->last - buf->pos;
+
+        if (size > 0) {
+          new_buf = ngx_create_temp_buf(&pool, size);
+          if (!new_buf) {
+            return NGX_ERROR;
+          }
+          new_buf->last = ngx_copy(new_buf->pos, buf->pos, size);
+          buf->pos = buf->last;  // consume
+          filter_ctx.copied_total += size;
+        } else {
+          // special buffer
+          if (!ngx_buf_special(buf)) {
+            ngx_log_error(
+                NGX_LOG_NOTICE, pool.log, 0,
+                "unexpected empty non-special buffer in buffer_chain");
+          }
+          new_buf = static_cast<ngx_buf_t *>(ngx_calloc_buf(&pool));
+          if (!new_buf) {
+            return NGX_ERROR;
+          }
+          new_buf->flush = buf->flush;
+          new_buf->sync = buf->sync;
+        }
+      } else {
+        // file buffers (or mixed memory/file buffers)
+        new_buf = static_cast<decltype(new_buf)>(ngx_calloc_buf(&pool));
+        if (!new_buf) {
+          return NGX_ERROR;
+        }
+        new_buf->in_file = 1;
+        new_buf->file = buf->file;
+        new_buf->file_pos = buf->file_pos;
+        new_buf->file_last = buf->file_last;
+
+        buf->file_pos = buf->file_last;  // consume
+
+        // mixed
+        if (buf->temporary) {
+          size = buf->last - buf->pos;
+          new_buf->temporary = 1;
+          if (size > 0) {
+            new_buf->pos = static_cast<u_char *>(ngx_palloc(&pool, size));
+            if (!new_buf->pos) {
+              return NGX_ERROR;
+            }
+            new_buf->last = ngx_copy(new_buf->pos, buf->pos, size);
+            buf->pos = buf->last;  // consume
+            filter_ctx.copied_total += size;
+          }
+        }
+
+        size = new_buf->file_last - new_buf->file_pos;
```

Comment:
If `buf->temporary` would size be any different than what was already calculated?

Reply:
I think mixed file/temporary buffers are the same size; at least that's what I saw. So yes, this line can be put in an else. (At the end, it doesn't really matter either way, as this size is only used for logging -- unlike `copied_total`)